client/client.cc

```diff
@@ -0,0 +1,195 @@
+using grpc::Channel;
+using grpc::ClientContext;
+using grpc::ClientReader;
+using grpc::Status;
+using std::pair;
+using std::vector;
+using supplyfinder::Finder;
+using supplyfinder::FinderRequest;
+using supplyfinder::FoodID;
+using supplyfinder::InventoryInfo;
+using supplyfinder::ShopInfo;
+using supplyfinder::Supplier;
+using supplyfinder::SupplierInfo;
+using supplyfinder::Vendor;
+using supplyfinder::VendorInfo;
+
+void PrintResult(const ShopInfo& p) {
+  std::cout << "\tVendor url: " << p.vendor().url()
+            << "; name: " << p.vendor().name()
+            << "; location: " << p.vendor().location() << std::endl;
+  std::cout << "\tInventory price: " << p.inventory().price()
+            << "; quantity: " << p.inventory().quantity() << "\n\n";
+}
+
+class FinderClient {
+ public:
+  FinderClient(std::shared_ptr<Channel> channel)
+      : stub_(Finder::NewStub(channel)) {}
+
+  void InquireFoodInfo(std::string& food_name, uint32_t quantity) {
+    FinderRequest request;
+    request.set_food_name(food_name);
+    request.set_quantity(quantity);
+    ShopInfo reply;
+    ClientContext context;
+    std::unique_ptr<ClientReader<ShopInfo>> reader(
+        stub_->CheckFood(&context, request));
+
+    std::cout << "Receiving Shop Information\n";
+    while (reader->Read(&reply)) {
+      PrintResult(reply);
+    }
+
+    Status status = reader->Finish();
+    if (!status.ok()) {
+      std::cout << status.error_code() << ": " << status.error_message()
+                << std::endl;
+    }
+  }
+
+  void UpdateSupplier(const std::string& addr) {
+    SupplierInfo info;
+    Empty empty;
+    ClientContext context;
+    info.set_url(addr);
+    Status status = stub_->UpdateSupplier(&context, info, &empty);
+    if (!status.ok()) {
+      std::cout << status.error_code() << ": " << status.error_message()
+                << std::endl;
+    }
+  }
+
+ private:
+  std::unique_ptr<Finder::Stub> stub_;
+};
+
+class Client {
+ public:
+  Client(std::string& finder_addr)
+      : finder_client_(grpc::CreateChannel(
+            finder_addr, grpc::InsecureChannelCredentials())) {}
+
+  void ProcessRequest(std::string& food_name, uint32_t quantity) {
+    finder_client_.InquireFoodInfo(food_name, quantity);
+  }
+
+  void InitSupplier(const std::string& supplier_addr) {
+    std::cout << "========== Initializing Supplier Server at " << supplier_addr
+              << " ==========" << std::endl;
+    finder_client_.UpdateSupplier(supplier_addr);
+
+    std::shared_ptr<Channel> channel =
+        grpc::CreateChannel(supplier_addr, grpc::InsecureChannelCredentials());
+    std::unique_ptr<Supplier::Stub> supplier_stub = Supplier::NewStub(channel);
+
+    VendorInfo info;
+    Empty empty;
+    std::string url, name, location;
+    std::cout
+        << "please input server url, name and location, separated by newlines.\n"
+        << "Or enter q to stop." << std::endl;
+    while (getline(std::cin, url) && url != "q") {
```

Comment:
I am pretty sure GKE has a way to provide addresses directly to where they are needed, e.i. finder would be provided the address and port of the supplier, supplier would be provided with the addresses of the vendors.  I don't know if you have enough time to look into this, but it's something to keep in mind.